KerbalStuff/blueprints/login_oauth.py

```diff
@@ -0,0 +1,329 @@
+        )
+    if 'error' in resp:
+        return jsonify(resp)
+    session['github_token'] = (resp['access_token'], '')
+    gh_info = github.get('user')
+    gh_info = gh_info.data
+    gh_user = gh_info['login']
+    auth = UserAuth.query.filter(
+            UserAuth.provider == 'github',
+            UserAuth.remote_user == gh_user).first()
+    if auth:
+        user = User.query.filter(User.id == auth.user_id).first()
+        if user.confirmation:
+            return redirect('/account-pending')
+        login_user(user, remember=True)
+        return redirect('/')
+    else:
+        emails = github.get('user/emails')
+        emails = emails.data
+        emails = [e['email'] for e in emails if e['primary']]
+        if emails:
+            email = emails[0]
+        else:
+            email = ''
+
+        return render_register_with_oauth('github', gh_user, gh_user, email)
+
+
+@login_oauth.route("/oauth/google/login")
+def login_with_oauth_authorized_google():
+    if 'code' not in request.args:
+        # Got here in some strange scenario.
+        return redirect('/')
+    google = get_oauth_provider('google')
+    resp = google.authorized_response()
+
+    if resp is None:
+        return 'Access denied: reason=%s error=%s' % (
+            request.args['error'],
+            request.args['error_description']
+        )
+    if 'error' in resp:
+        return jsonify(resp.error)
+    session['google_token'] = (resp['access_token'], '')
+    google_info = google.get('userinfo')
+    google_info = google_info.data
+    google_user = google_info['id']  # This is a long number.
+    auth = UserAuth.query.filter(
+            UserAuth.provider == 'google',
+            UserAuth.remote_user == google_user).first()
+    if auth:
+        user = User.query.filter(User.id == auth.user_id).first()
+        if user.confirmation:
+            return redirect('/account-pending')
+        login_user(user, remember=True)
+        return redirect('/')
+    else:
+        email = google_info['email']
+        username = email[:email.find('@')]
+
+        return render_register_with_oauth('google', google_user, username, email)
+
+
+@login_oauth.route("/register-oauth", methods=['POST'])
+def register_with_oauth_authorized():
+    '''
+    This endpoint should be called after authorizing with oauth, by the user.
+    '''
+    email = request.form.get('email')
+    username = request.form.get('username')
+    provider = request.form.get('provider')
+    remote_user = request.form.get('remote_user')
+
+    good = True
+    if check_username_for_registration(username):
+        good = False
+    if check_email_for_registration(email):
+        good = False
+
+    if good:
+        password = binascii.b2a_hex(os.urandom(99))
+        user = User(username, email, password)
+        user.confirmation = binascii.b2a_hex(os.urandom(20)).decode("utf-8")
+        db.add(user)
+        db.commit()  # Need to commit now to get an ID.
```

Comment:
When you use commit, you actually create rows in the database. When you flush, you just flush the transaction, and SQL gives you the ID, but doesn't commit the rows. If something goes wrong after a flush, there isn't going to be bad/incomplete data in the database.